onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -238,6 +241,36 @@ class PlannerImpl {
     return false;
   }
 
+  bool FindReusableGraphInput(const onnxruntime::Node& node, int output_arg_num, MLValueIndex* reusable_input) {
+    // skip if this is main graph
+    if (parent_node_) {
+      // only check a single layer. ie. an Identity/Dropout node connecting graph's input and output
+      auto& graph_inputs = graph_viewer_.GetInputs();
+      auto& graph_outputs = graph_viewer_.GetOutputs();
+
+      if (FindReusableInput(node, output_arg_num, reusable_input)) {
```

Comment:
Okay, after looking through this again, there are still some edge cases where this can be a problem. First, FindReusableInput looks for opportunities for both Alias (identity copy operations) and MayInplace (where we use destructive in-place updates). So, for example, if we have X_out = ADD (X_in, Y), and "ADD" has MayInplace, this optimization will kick in. Do we want that? On one hand, it seems useful when it is valid. On the other hand, there are some cases where it is invalid. E.g., if the initial-value of "X" is some constant tensor C defined in an initializer, this will end up overwriting that constant tensor C. So, at the least, it looks like some extra checks will be needed.